blockchain/src/state/read/convert.rs

```diff
@@ -0,0 +1,178 @@
+        &self,
+        layout: &MoveTypeLayout,
+        value: serde_json::Value,
+    ) -> Result<MoveValue> {
+        let value = match layout {
+            MoveTypeLayout::Bool => {
+                MoveValue::Bool(serde_json::from_value(value).map_err(incorrect_type_layout)?)
+            }
+            MoveTypeLayout::U8 => {
+                MoveValue::U8(serde_json::from_value(value).map_err(incorrect_type_layout)?)
+            }
+            MoveTypeLayout::U16 => {
+                MoveValue::U16(serde_json::from_value(value).map_err(incorrect_type_layout)?)
+            }
+            MoveTypeLayout::U32 => {
+                MoveValue::U32(serde_json::from_value(value).map_err(incorrect_type_layout)?)
+            }
+            MoveTypeLayout::U64 => serde_json::from_value::<U64>(value)
+                .map_err(incorrect_type_layout)?
+                .into(),
+            MoveTypeLayout::U128 => serde_json::from_value::<U128>(value)
+                .map_err(incorrect_type_layout)?
+                .into(),
+            MoveTypeLayout::U256 => serde_json::from_value::<U256>(value)
+                .map_err(incorrect_type_layout)?
+                .into(),
+            MoveTypeLayout::Address => serde_json::from_value::<Address>(value)
+                .map_err(incorrect_type_layout)?
+                .into(),
+            MoveTypeLayout::Vector(item_layout) => {
+                self.try_into_vm_value_vector(item_layout, value)?
+            }
+            MoveTypeLayout::Struct(struct_layout) => {
+                self.try_into_vm_value_struct(struct_layout, value)?
+            }
+            MoveTypeLayout::Signer | MoveTypeLayout::Native(..) => {
+                // Signer and Native types are not supported
+                return Err(UserError::IncorrectTypeLayout.into());
+            }
+        };
+        Ok(value)
+    }
+
+    pub fn try_into_vm_value_vector(
+        &self,
+        layout: &MoveTypeLayout,
+        value: serde_json::Value,
+    ) -> Result<MoveValue> {
+        if matches!(layout, MoveTypeLayout::U8) {
+            Ok(serde_json::from_value::<HexEncodedBytes>(value)
+                .map_err(incorrect_type_layout)?
+                .into())
+        } else if let serde_json::Value::Array(list) = value {
+            let vals = list
+                .into_iter()
+                .map(|v| self.try_into_vm_value_from_layout(layout, v))
+                .collect::<Result<_>>()?;
+
+            Ok(MoveValue::Vector(vals))
+        } else {
+            Err(incorrect_type_layout(()).into())
+        }
+    }
+
+    pub fn try_into_vm_value_struct(
+        &self,
+        layout: &MoveStructLayout,
+        value: serde_json::Value,
+    ) -> Result<MoveValue> {
+        let (struct_tag, field_layouts) =
+            if let MoveStructLayout::WithTypes { type_, fields } = layout {
+                (type_, fields)
+            } else {
+                return Err(incorrect_type_layout(()).into());
```

Comment:
Is a wrapping function for the error here just serving brevity? I feel it would be more straightforward to just use the variant itself.

Reply:
I was doing it for consistency with the other error handling (`map_err` on results). But I agree it is less clear than just using the variant directly. Fixed in https://github.com/UmiNetwork/op-move/pull/502/commits/1ea77e5332e749fa461ce3a91bbcb402f65592af